### PR TITLE
[WIP] Implemented os_scandir

### PR DESCRIPTION
**Pull request by jonaias**
Continuing my previous work(#2010),  this PR enables cross platform support for expand path. It uses libuv scandir. 

Edit:
@equalsraf  I was able to remove POSIX "dirent.h" dependency, easier to port =)

### CODE REVIEW

File: src/nvim/os/fs.c [hunk 1]
```diff
@@ -503,3 +503,42 @@ bool os_fileid_equal_fileinfo(const FileID *file_id,
          && file_id->device_id == file_info->stat.st_dev;
 }
 
+
+/// Open directory for listing contents
+///
+/// @param scandir_req scandir request
+/// @param directory_path Directory path to be listed
+/// @return The number of directory entries
+int os_scandir_open(uv_fs_t *scandir_req, const char *directory_path)
```
Comment (fwalch) [inline src/nvim/os/fs.c]: I think we should not expose LibUV types, what about doing something like `FileInfo` in `fs_defs.h`?

**Comment by equalsraf** [comment]
Hi

@jonaias  have you seen #1212?. There is some clear overlap with this PR.

@justinmk any idea what is the status for USE_FNAME_CASE and #1212? If USE_FNAME_CASE is set I think this wont build because of dirent/dp in `fname_case`. Some of the issues in #1212 should be solved upstream by now, either there or here I think this can move forward by now.

File: src/nvim/os/fs.c (hunk 1 above)
Comment (jonaias) [inline src/nvim/os/fs.c]: You're right!

**Comment by jonaias** [comment]
@equalsraf  I didn't. Thanks for the heads up. I'll find something else to work.

@splinterofchaos I'll close this PR and will pay attention in yours PR, until it gets merged.

**Comment by splinterofchaos** [comment]
What a coincidence! I've been busy for a while, but I was just about to revive that PR. I just rebased and pushed the changes. It still needs some Mac-specific unit tests for completeness, though. When I wrote #1212, travis didn't seem to work on Mac so `path_fix_case` couldn't be tested.

**Pull request closed without merging**